python/lsst/daf/butler/_butler.py

```diff
@@ -1597,6 +1599,161 @@ def retrieveArtifacts(
             overwrite=overwrite,
         )
 
+    def exists(
+        self,
+        dataset_ref_or_type: DatasetRef | DatasetType | str,
+        /,
+        data_id: DataId | None = None,
+        *,
+        full_check: bool = True,
+        collections: Any = None,
+        **kwargs: Any,
+    ) -> DatasetExistence:
+        """Indicate whether a dataset is known to Butler registry and
+        datastore.
+
+        Parameters
+        ----------
+        dataset_ref_or_type : `DatasetRef`, `DatasetType`, or `str`
+            When `DatasetRef` the `dataId` should be `None`.
+            Otherwise the `DatasetType` or name thereof.
+        data_id : `dict` or `DataCoordinate`
+            A `dict` of `Dimension` link name, value pairs that label the
+            `DatasetRef` within a Collection. When `None`, a `DatasetRef`
+            should be provided as the first argument.
+        full_check : `bool`, optional
+            If `True`, an additional check will be made for dataset artifact
+            existence. This will involve additional overhead due to the need
+            to query an external system. If `False` registry and datastore
+            will solely be asked if they know about the dataset but no
+            check for the artifact will be performed.
+        collections : Any, optional
+            Collections to be searched, overriding ``self.collections``.
+            Can be any of the types supported by the ``collections`` argument
+            to butler construction.
```

Comment:
One more thought - it looks like `collections` is not used when `ref` is given, meaning that we'll find a ref even if it's not in any of the `collections`. Maybe clarify that in the docstring to avoid surprises for users who may think that `collections` could be used as restriction even if dataset ref is provided?

Reply:
Right. We do warn in `Butler.get()` now if you pass ref and collections.